core/src/test/scala/unit/kafka/server/KafkaApisTest.scala

```diff
@@ -363,6 +363,53 @@ class KafkaApisTest {
     testConsumerListOffsetLatest(IsolationLevel.READ_COMMITTED)
   }
 
+  /**
+   * Verifies that the metadata response is correct if the broker listeners are inconsistent (i.e. one broker has
+   * more listeners than another) and the request is sent on the listener that exists in both brokers.
+   */
+  @Test
+  def testMetadataRequestOnSharedListenerWithInconsistentListenersAcrossBrokers(): Unit = {
+    val response = sendMetadataRequestWithInconsistentListeners(ListenerName.forSecurityProtocol(SecurityProtocol.PLAINTEXT))
+    assertEquals(Set(0, 1), response.brokers.asScala.map(_.id).toSet)
+  }
+
+  /*
+   * Verifies that the metadata response is correct if the broker listeners are inconsistent (i.e. one broker has
+   * more listeners than another) and the request is sent on the listener that exists in one broker.
+   */
+  @Test
+  def testMetadataRequestOnDistinctListenerWithInconsistentListenersAcrossBrokers(): Unit = {
+    val response = sendMetadataRequestWithInconsistentListeners(new ListenerName("LISTENER2"))
+    assertEquals(Set(0), response.brokers.asScala.map(_.id).toSet)
+  }
+
+  private def sendMetadataRequestWithInconsistentListeners(requestListener: ListenerName): MetadataResponse = {
```

Comment:
I'd suggest turning this into two functions: one which updates the metadata cache with the inconsistent listener names, and another which sends the metadata request. Then you can make the dependence on the lister names explicit.